File: Servers/HeadServer/HeadServer.cs
﻿using System.Collections;
using System.Collections.Generic;
using CommonShuffleLibraries;
using NodeJSLibrary;

namespace HeadServer
{
    public class HeadServer
    {
        private FS fs = Global.Require<FS>("fs");

        private string __dirname = "/usr/local/src/headServer";
        private string indexPageData;
        private QueueManager qManager;
        private PubSub pubsub;

        private List<string> indexForSites = new List<string>();
        private List<string> gateways = new List<string>();
        private List<string> oldGateways = new List<string>();
        private List<string> oldIndex = new List<string>();
        private int siteIndex = 0;

        public HeadServer()
        {
            qManager = new QueueManager("Head1", new QueueManagerOptions(new QueueWatcher[]
                                                                             {
                                                                                 new QueueWatcher("HeadServer", null),
                                                                                 new QueueWatcher("Head1", null),
                                                                             }, new string[]
                                                                                    {
                                                                                        "GatewayServer"
                                                                                    }));

            
            fs.ReadFile(__dirname + "/index.html", ready);

            pubsub = new PubSub(delegate
                                    {
                                        pubsub.Subscribe("PUBSUB.GatewayServers", delegate(object message)
                                                                                      {
                                                                                          indexForSites.Add(indexPageData.Replace("{{gateway}}", message.ToString()));
                                                                                          gateways.Add(message.ToString());
                                                                                      });
                                    });

            Global.Require<Http>("http").CreateServer(handlerWS).Listen(8844);
            qManager.AddChannel<string>("Head.GatewayUpdate", delegate(User user, object data)
                                                          {
                                                              indexForSites.Add(indexPageData.Replace("{{gateway}}", data.ToString()));
                                                              gateways.Add(data.ToString());
                                                          });

            Global.SetInterval(pollGateways, 5000);
            pollGateways();
        }

        private void pollGateways()
        {
            pubsub.Publish("PUBSUB.GatewayServers.Ping", "");

            if(indexForSites.Count>0)
            {
                oldIndex = indexForSites;
            }
            if(gateways.Count>0)
            {
                oldGateways = gateways;
            }
            indexForSites = new List<string>();
            gateways = new List<string>();
            siteIndex = 0;
        }

        private void handlerWS(HttpRequest request, HttpResponse response)
        {
            if (oldGateways.Count > 0)
            {
                   int inj = (siteIndex++) % oldIndex.Count;
                response.End(oldGateways[inj]);
                return;
            }
            response.End();
        }

        private void handler(HttpRequest request, HttpResponse response)
        {
            JsDictionary dict = new JsDictionary();
            dict["Content-Type"] = "text/html";
            if(oldIndex.Count>0)
            {
                response.WriteHead(200, dict);
                int inj = (siteIndex++)%oldIndex.Count;
                response.End(oldIndex[inj]);
            }
            else
            {


                response.WriteHead(200, dict);
                response.End();
            }
        }
        public void ready(FileSystemError error,object content)
        {
            indexPageData = content.ToString();
            Global.Require<Http>("http").CreateServer(handler).Listen(80);

        }

    }

}

File: output/HeadServer.js
require('./mscorlib.js');require('./CommonShuffleLibrary.js');require('./Models.js');
////////////////////////////////////////////////////////////////////////////////
// HeadServer.HeadServer
var $HeadServer_HeadServer = function() {
	this.$__dirname = '/usr/local/src/new';
	this.$fs = require('fs');
	this.$gateways = [];
	this.$indexForSites = [];
	this.$indexPageData = null;
	this.$oldGateways = [];
	this.$oldIndex = [];
	this.$pubsub = null;
	this.$qManager = null;
	this.$siteIndex = 0;
	this.$qManager = new CommonShuffleLibrary.QueueManager('Head1', new CommonShuffleLibrary.QueueManagerOptions([new CommonShuffleLibrary.QueueWatcher('HeadServer', null), new CommonShuffleLibrary.QueueWatcher('Head1', null)], ['GatewayServer']));
	this.$fs.readFile(this.$__dirname + '/index.html', 'ascii', Function.mkdel(this, this.ready));
	this.$pubsub = new CommonShuffleLibrary.PubSub(Function.mkdel(this, function() {
		this.$pubsub.subscribe('PUBSUB.GatewayServers', Function.mkdel(this, function(message) {
			this.$indexForSites.add(this.$indexPageData.replaceAll('{{gateway}}', message));
			this.$gateways.add(message);
		}));
	}));
	require('http').createServer(Function.mkdel(this, this.$handlerWS)).listen(8844);
	setInterval(Function.mkdel(this, this.$pollGateways), 1000);
	this.$pollGateways();
};
$HeadServer_HeadServer.prototype = {
	$pollGateways: function() {
		this.$pubsub.publish('PUBSUB.GatewayServers.Ping', '');
		if (this.$indexForSites.length > 0) {
			this.$oldIndex = this.$indexForSites;
		}
		if (this.$gateways.length > 0) {
			this.$oldGateways = this.$gateways;
		}
		this.$indexForSites = [];
		this.$gateways = [];
		this.$siteIndex = 0;
	},
	$handlerWS: function(request, response) {
		if (this.$oldGateways.length > 0) {
			var inj = this.$siteIndex++ % this.$oldIndex.length;
			response.end(this.$oldGateways[inj]);
			return;
		}
		response.end();
	},
	$handler: function(request, response) {
		var dict = {};
		dict['Content-Type'] = 'text/html';
		if (this.$oldIndex.length > 0) {
			response.writeHead(200, dict);
			var inj = this.$siteIndex++ % this.$oldIndex.length;
			response.end(this.$oldIndex[inj]);
		}
		else {
			response.writeHead(200, dict);
			response.end();
		}
	},
	ready: function(error, content) {
		this.$indexPageData = content.toString();
		require('http').createServer(Function.mkdel(this, this.$handler)).listen(80);
	}
};
$HeadServer_HeadServer.main = function() {
	try {
		new $HeadServer_HeadServer();
	}
	catch ($t1) {
		var exc = ss.Exception.wrap($t1);
		console.log('CRITICAL FAILURE: ' + CommonLibraries.ExtensionMethods.goodMessage(exc));
	}
};
Type.registerClass(global, 'HeadServer.HeadServer', $HeadServer_HeadServer, Object);
$HeadServer_HeadServer.main();
